controllers/snapshot/snapshot_adapter.go

```diff
@@ -75,6 +76,87 @@ func scenariosNamesToList(integrationTestScenarions *[]v1beta1.IntegrationTestSc
 	return &result
 }
 
+// EnsureRerunPipelineRunsExist is responsible for recreating integration test pipelines triggered by users
+func (a *Adapter) EnsureRerunPipelineRunsExist() (controller.OperationResult, error) {
+
+	scenarioName, ok := gitops.GetIntegrationTestRunLabelValue(*a.snapshot)
+	if !ok {
+		// no test rerun trigerred
+		return controller.ContinueProcessing()
+	}
+	integrationTestScenario, err := a.loader.GetScenario(a.client, a.context, scenarioName, a.application.Namespace)
+
+	if err != nil {
+		if clienterrors.IsNotFound(err) {
+			a.logger.Error(err, "scenario for integration test re-run not found", "scenario", scenarioName)
+			// scenario doesn't exist just remove label and continue
+			if err = gitops.RemoveIntegrationTestRerunLabel(a.client, a.context, a.snapshot); err != nil {
+				return controller.RequeueWithError(err)
+			}
+			return controller.ContinueProcessing()
+		}
+		return controller.RequeueWithError(fmt.Errorf("failed to fetch requested scenario %s: %w", scenarioName, err))
+	}
+
+	a.logger.Info("Re-running integration test for scenario", "scenario", scenarioName)
+
+	testStatuses, err := gitops.NewSnapshotIntegrationTestStatusesFromSnapshot(a.snapshot)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	testStatuses.ResetStatus(scenarioName)
+
+	if shouldScenarioRunInEphemeralEnv(integrationTestScenario) {
+		allEnvironments, err := a.loader.GetAllEnvironments(a.client, a.context, a.application)
+		if err != nil {
+			a.logger.Error(err, "Failed to get all environments.")
+			return controller.RequeueWithError(err)
+		}
+
+		_, err = a.ensureEphemeralEnvironmentForScenarioExists(integrationTestScenario, testStatuses, allEnvironments)
+		if err != nil {
+			if h.IsEnvironmentNotInNamespaceError(err) {
+				a.logger.Error(err, "environment doesn't exist in the same namespace as integrationScenario at all, try again after creating environment",
+					"integrationTestScenario.Name", integrationTestScenario.Name)
+				// remove label to prevent indefinite failures, users must restart test manually again
+				if err = gitops.RemoveIntegrationTestRerunLabel(a.client, a.context, a.snapshot); err != nil {
+					return controller.RequeueWithError(err)
+				}
+				return controller.StopProcessing()
+			}
+			return controller.RequeueWithError(fmt.Errorf("failed to ensure existence of the environment for scenario %s: %w", integrationTestScenario.Name, err))
+		}
+
+	} else {
+		integrationTestScenarioStatus, ok := testStatuses.GetScenarioStatus(integrationTestScenario.Name)
+		if ok && integrationTestScenarioStatus.TestPipelineRunName != "" {
```

Comment:
Do we need this check here since the status for that scenario will be reset `testStatuses.ResetStatus(scenarioName)` at this point?

Reply:
yes, in case that for unknown reason, label failed to be removed and reconciliation may re-trigger another test.

Reply:
I see your point, that's certainly possible, but wouldn't the `TestPipelineRunName` always be removed [here](https://github.com/redhat-appstudio/integration-service/pull/408/files#diff-eeed1ae6f738d926a3e659cd3b10bbeb1f2c2a05f7ab22d899fa6914059660c0R122) before we check this?

Reply:
yes /o\

Reply:
That's a tough case actually. What if we don't allow rerunning Integration tests that are still running?

Reply:
Updated and unittest added